File: android/src/main/java/com/resquircle/SquirclePath.kt
package com.resquircle

import android.graphics.Path
import android.graphics.RectF
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt
import kotlin.math.tan

data class CurveProperties(
  var a: Float,
  var b: Float,
  var c: Float,
  var d: Float,
  var p: Float,
  var arcSectionLength: Float,
  var cornerRadius: Float,
)

class SquirclePath(
  private var width: Float,
  private var height: Float,
  private var borderRadius: Float,
  private var cornerSmoothing: Float,
) : Path() {

  init {
    val checkedRadius = minOf(this.borderRadius, this.width / 2f, this.height / 2f)
    val checkedCornerSmoothing = maxOf(minOf(this.cornerSmoothing, 1f), 0f)

    if (checkedRadius <= 0f || width <= 0f || height <= 0f) {
      addRect(0f, 0f, width, height, Direction.CW)
    } else {
      val curvedProperties =
        calculateCurveProperties(
          checkedRadius,
          checkedCornerSmoothing,
          min(this.width, this.height) / 2,
        )

      buildPath(this.width, this.height, curvedProperties)
    }
  }

  private fun buildPath(width: Float, height: Float, cp: CurveProperties) {
    val a = cp.a
    val b = cp.b
    val c = cp.c
    val d = cp.d
    val p = cp.p
    val arc = cp.arcSectionLength
    val r = cp.cornerRadius

    val topRightCx = width - r
    val topRightCy = r
    val bottomRightCx = width - r
    val bottomRightCy = height - r
    val bottomLeftCx = r
    val bottomLeftCy = height - r
    val topLeftCx = r
    val topLeftCy = r

    var x = width - p
    var y = 0f
    moveTo(x, y)

    fun relCubic(c1x: Float, c1y: Float, c2x: Float, c2y: Float, dx: Float, dy: Float) {
      rCubicTo(c1x, c1y, c2x, c2y, dx, dy)
      x += dx
      y += dy
    }

    fun arcToRelative(deltaX: Float, deltaY: Float, centerX: Float, centerY: Float, radius: Float) {
      val endX = x + deltaX
      val endY = y + deltaY

      val startAngle =
        Math.toDegrees(kotlin.math.atan2((y - centerY), (x - centerX)).toDouble()).toFloat()
      val endAngle =
        Math.toDegrees(kotlin.math.atan2((endY - centerY), (endX - centerX)).toDouble()).toFloat()

      var sweep = endAngle - startAngle
      if (sweep < 0f) sweep += 360f

      val oval = RectF(centerX - radius, centerY - radius, centerX + radius, centerY + radius)
      arcTo(oval, startAngle, sweep, false)

      x = endX
      y = endY
    }

    relCubic(a, 0f, a + b, 0f, a + b + c, d)
    arcToRelative(arc, arc, topRightCx, topRightCy, r)
    relCubic(d, c, d, c + d, d, a + b + c)

    x = width
    y = height - p
    lineTo(x, y)

    relCubic(0f, a, 0f, a + b, -d, a + b + c)
    arcToRelative(-arc, arc, bottomRightCx, bottomRightCy, r)
    relCubic(-c, d, -(b + c), d, -(a + b + c), d)

    x = p
    y = height
    lineTo(x, y)

    relCubic(-a, 0f, -(a + b), 0f, -(a + b + c), -d)
    arcToRelative(-arc, -arc, bottomLeftCx, bottomLeftCy, r)
    relCubic(-d, -c, -d, -(b + c), -d, -(a + b + c))

    x = 0f
    y = p
    lineTo(x, y)

    relCubic(0f, -a, 0f, -(a + b), d, -(a + b + c))
    arcToRelative(arc, -arc, topLeftCx, topLeftCy, r)
    relCubic(c, -d, b + c, -d, a + b + c, -d)

    close()
  }

  private fun calculateCurveProperties(
    cornerRadius: Float,
    cornerSmoothing: Float,
    roundingAndSmoothingBudget: Float,
  ): CurveProperties {
    var p = (1 + cornerSmoothing) * cornerRadius

    val arcMeasure = 90 * (1 - cornerSmoothing)
    val arcSectionLength = sin(toRadians(arcMeasure / 2)) * cornerRadius * sqrt(2f)
    val angleAlpha = (90 - arcMeasure) / 2
    val p3ToP4Distance = cornerRadius * tan(toRadians(angleAlpha / 2))
    val angleBeta = 45 * cornerSmoothing
    val c = p3ToP4Distance * cos(toRadians(angleBeta))
    val d = c * tan(toRadians(angleBeta))
    var b = (p - arcSectionLength - c - d) / 3
    var a = 2 * b

    if (p > roundingAndSmoothingBudget) {
      val p1ToP3MaxDistance = roundingAndSmoothingBudget - d - arcSectionLength - c
      val minA = p1ToP3MaxDistance / 6
      val maxB = p1ToP3MaxDistance - minA
      b = minOf(b, maxB)
      a = p1ToP3MaxDistance - b
      p = minOf(p, roundingAndSmoothingBudget)
    }

    return CurveProperties(a, b, c, d, p, arcSectionLength, cornerRadius)
  }

  private fun toRadians(degrees: Float): Float {
    return degrees * (Math.PI.toFloat() / 180f)
  }
}

File: ios/SquirclePath.swift
//
//  SquirclePath.swift
//  Ported from expo-squircle-view (MIT)
//

import UIKit

struct CurveProperties {
  var a: CGFloat
  var b: CGFloat
  var c: CGFloat
  var d: CGFloat
  var p: CGFloat
  var arcSectionLength: CGFloat
  var cornerRadius: CGFloat
}

struct SquirclePath {
  static func create(
    width: CGFloat,
    height: CGFloat,
    radius: CGFloat,
    cornerSmoothing: CGFloat
  ) -> CGPath {
    if width <= 0 || height <= 0 {
      return CGPath(rect: .zero, transform: nil)
    }

    let checkedRadius = min(radius, width / 2, height / 2)
    let checkedCornerSmoothing = max(min(cornerSmoothing, 1), 0)

    let curveProperties = calculateCurveProperties(
      cornerRadius: checkedRadius,
      cornerSmoothing: checkedCornerSmoothing,
      roundingAndSmoothingBudget: min(width, height) / 2
    )

    return createPath(width: width, height: height, curveProperties: curveProperties)
  }

  private static func createPath(
    width: CGFloat,
    height: CGFloat,
    curveProperties cp: CurveProperties
  ) -> CGPath {
    if cp.cornerRadius <= 0 {
      return CGPath(rect: CGRect(x: 0, y: 0, width: width, height: height), transform: nil)
    }

    let a = cp.a
    let b = cp.b
    let c = cp.c
    let d = cp.d
    let p = cp.p
    let arc = cp.arcSectionLength
    let r = cp.cornerRadius

    let topRightCenter = CGPoint(x: width - r, y: r)
    let bottomRightCenter = CGPoint(x: width - r, y: height - r)
    let bottomLeftCenter = CGPoint(x: r, y: height - r)
    let topLeftCenter = CGPoint(x: r, y: r)

    let path = UIBezierPath()

    var current = CGPoint(x: width - p, y: 0)
    path.move(to: current)

    // Top-right corner
    current = addRelativeCurve(
      path,
      from: current,
      c1: CGPoint(x: a, y: 0),
      c2: CGPoint(x: a + b, y: 0),
      end: CGPoint(x: a + b + c, y: d)
    )
    current = addArc(path, from: current, by: CGPoint(x: arc, y: arc), center: topRightCenter, radius: r)
    current = addRelativeCurve(
      path,
      from: current,
      c1: CGPoint(x: d, y: c),
      c2: CGPoint(x: d, y: c + d),
      end: CGPoint(x: d, y: a + b + c)
    )

    // Right edge
    current = CGPoint(x: width, y: height - p)
    path.addLine(to: current)

    // Bottom-right corner
    current = addRelativeCurve(
      path,
      from: current,
      c1: CGPoint(x: 0, y: a),
      c2: CGPoint(x: 0, y: a + b),
      end: CGPoint(x: -d, y: a + b + c)
    )
    current = addArc(path, from: current, by: CGPoint(x: -arc, y: arc), center: bottomRightCenter, radius: r)
    current = addRelativeCurve(
      path,
      from: current,
      c1: CGPoint(x: -c, y: d),
      c2: CGPoint(x: -(b + c), y: d),
      end: CGPoint(x: -(a + b + c), y: d)
    )

    // Bottom edge
    current = CGPoint(x: p, y: height)
    path.addLine(to: current)

    // Bottom-left corner
    current = addRelativeCurve(
      path,
      from: current,
      c1: CGPoint(x: -a, y: 0),
      c2: CGPoint(x: -(a + b), y: 0),
      end: CGPoint(x: -(a + b + c), y: -d)
    )
    current = addArc(path, from: current, by: CGPoint(x: -arc, y: -arc), center: bottomLeftCenter, radius: r)
    current = addRelativeCurve(
      path,
      from: current,
      c1: CGPoint(x: -d, y: -c),
      c2: CGPoint(x: -d, y: -(b + c)),
      end: CGPoint(x: -d, y: -(a + b + c))
    )

    // Left edge
    current = CGPoint(x: 0, y: p)
    path.addLine(to: current)

    // Top-left corner
    current = addRelativeCurve(
      path,
      from: current,
      c1: CGPoint(x: 0, y: -a),
      c2: CGPoint(x: 0, y: -(a + b)),
      end: CGPoint(x: d, y: -(a + b + c))
    )
    current = addArc(path, from: current, by: CGPoint(x: arc, y: -arc), center: topLeftCenter, radius: r)
    _ = addRelativeCurve(
      path,
      from: current,
      c1: CGPoint(x: c, y: -d),
      c2: CGPoint(x: b + c, y: -d),
      end: CGPoint(x: a + b + c, y: -d)
    )

    path.close()
    return path.cgPath
  }

  private static func addRelativeCurve(
    _ path: UIBezierPath,
    from current: CGPoint,
    c1: CGPoint,
    c2: CGPoint,
    end: CGPoint
  ) -> CGPoint {
    let control1 = CGPoint(x: current.x + c1.x, y: current.y + c1.y)
    let control2 = CGPoint(x: current.x + c2.x, y: current.y + c2.y)
    let endPoint = CGPoint(x: current.x + end.x, y: current.y + end.y)
    path.addCurve(to: endPoint, controlPoint1: control1, controlPoint2: control2)
    return endPoint
  }

  private static func addArc(
    _ path: UIBezierPath,
    from current: CGPoint,
    by delta: CGPoint,
    center: CGPoint,
    radius: CGFloat
  ) -> CGPoint {
    let endPoint = CGPoint(x: current.x + delta.x, y: current.y + delta.y)
    let startAngle = atan2(current.y - center.y, current.x - center.x)
    let endAngle = atan2(endPoint.y - center.y, endPoint.x - center.x)
    path.addArc(withCenter: center, radius: radius, startAngle: startAngle, endAngle: endAngle, clockwise: true)
    return endPoint
  }

  static func calculateCurveProperties(
    cornerRadius: CGFloat,
    cornerSmoothing: CGFloat,
    roundingAndSmoothingBudget: CGFloat
  ) -> CurveProperties {
    var p = (1 + cornerSmoothing) * cornerRadius

    let arcMeasure = 90 * (1 - cornerSmoothing)
    let arcSectionLength = sin(toRadians(arcMeasure / 2)) * cornerRadius * sqrt(2)
    let angleAlpha = (90 - arcMeasure) / 2
    let p3ToP4Distance = cornerRadius * tan(toRadians(angleAlpha / 2))
    let angleBeta = 45 * cornerSmoothing
    let c = p3ToP4Distance * cos(toRadians(angleBeta))
    let d = c * tan(toRadians(angleBeta))
    var b = (p - arcSectionLength - c - d) / 3
    var a = 2 * b

    if p > roundingAndSmoothingBudget {
      let p1ToP3MaxDistance = roundingAndSmoothingBudget - d - arcSectionLength - c
      let minA = p1ToP3MaxDistance / 6
      let maxB = p1ToP3MaxDistance - minA
      b = min(b, maxB)
      a = p1ToP3MaxDistance - b
      p = min(p, roundingAndSmoothingBudget)
    }

    return CurveProperties(
      a: a,
      b: b,
      c: c,
      d: d,
      p: p,
      arcSectionLength: arcSectionLength,
      cornerRadius: cornerRadius
    )
  }

  static func toRadians(_ degrees: CGFloat) -> CGFloat {
    degrees * .pi / 180
  }
}

